src/jit/gcencode.cpp

```diff
@@ -4204,9 +4221,26 @@ void GCInfo::gcMakeRegPtrTable(GcInfoEncoder* gcInfoEncoder,
         {
             if (gcCallDescList != nullptr)
             {
-                for (CallDsc* call = gcCallDescList; call != nullptr; call = call->cdNext)
+#if defined(_TARGET_AMD64_) && defined(FEATURE_CORECLR)
+                if (fastPath)
                 {
-                    numCallSites++;
+                    // We have the call count from the previous run.
+                    numCallSites = *callCntRef;
+
+                    // If there are no calls, tell the world and bail.
+                    if (numCallSites == 0)
+                    {
+                        gcInfoEncoderWithLog->DefineCallSites(nullptr, nullptr, 0);
+                        return;
+                    }
```

Comment:
The main goal is to skip all the code below here because we know (from the first pass) that we'll find no calls that we'll need to record, but there are usually quite a few calls on the list (they just all get skipped twice in 99.9% of the cases). The only way to avoid wrapping everything below in yet another level of nesting would be to bash the call list to null but that seems extremely dirty (and may not work). If you have a specific idea, though, please let me know - I agree the method is pretty knarly as it is.